### ISSUE TITLE
Dynamic "Download as..." options

### ISSUE THREAD

**Issue by janfreyberg** [jupyter/notebook#2177]
Since nbconvert allows custom exporters to be added to entry points, it would be neat if the notebook GUI recognised if custom exporters showed up in the `File > Download as..` menu for files with the right file type.

If eg a file is opened with a MATLAB kernel, which specifies `.m` file extensions, and there is a custom `.m` exporter installed, that export option would appear in the menu.

**Comment by blink1073** [comment]
The notebook is populating that list using an http request that queries `nbconvert` directly: https://github.com/jupyter/notebook/blob/b2b12eca955d84b9112510ca041a057cb9472a84/notebook/services/nbconvert/handlers.py#L13

I think the issue is that the [exporter map](https://github.com/jupyter/nbconvert/blob/605bac326cb913a437ff56a38ba13152a7ff2364/nbconvert/exporters/exporter_locator.py#L39).  
`matlab_nbconvert` is using an entry point to register: https://github.com/janfreyberg/matlab_nbconvert/blob/ad70b708b567f9fb53d77b65e11a178daf9883be/setup.py#L39.

Aha! The import used in that handler is deprecated: https://github.com/jupyter/nbconvert/blob/master/nbconvert/exporters/export.py

It should be using: https://github.com/jupyter/nbconvert/blob/605bac326cb913a437ff56a38ba13152a7ff2364/nbconvert/exporters/base.py#L113

@janfreyberg, you should also restore the mimetype in your exporter since it is used in the handler.

**Comment by takluyver** [comment]
Hmm, I thought that list was hardcoded in the notebook. @janfreyberg asked about this on nbconvert and I asked him to open an issue here.

**Comment by janfreyberg** [comment]
This might also require custom exporters to have a minimum amount of metadata, like a clean name? My custom exporter doesn't specify anything beyond mime type and file extension, which doesn't seem like enough to create a proper menu entry.

**Comment by blink1073** [comment]
Ah, I was looking at the REST endpoint.  The frontend is indeed hard-coded: https://github.com/jupyter/notebook/blob/8fbf1fb1ce1fa1d219d6b86deb27a0b1a3e65703/notebook/static/notebook/js/menubar.js#L193.

**Comment by blink1073** [comment]
In the template: https://github.com/jupyter/notebook/blob/8fbf1fb1ce1fa1d219d6b86deb27a0b1a3e65703/notebook/templates/notebook.html#L105.

**Comment by blink1073** [comment]
To make this dynamic, the notebook file handler would have to populate that list from the known nbconvert templates, but `TemplateExporter` class would have to gain a trait for its display name.

@michaelpacer, thoughts?

**Comment by blink1073** [comment]
Ah, it seems we already get that:

```python
In [2]: from nbconvert import exporters

In [3]: exporters.get_export_names()
Out[3]:
['asciidoc',
 'custom',
 'html',
 'html_embed',
 'html_toc',
 'html_with_lenvs',
 'html_with_toclenvs',
 'latex',
 'latex_with_lenvs',
 'markdown',
 'matlab',
 'notebook',
 'pdf',
 'python',
 'rst',
 'script',
 'slides']
```

**Comment by takluyver** [comment]
Looking back at your use case, it might be useful to mention that we're planning to provide a way to override the 'script' exporter in nbconvert to do something more specific for particular languages. This might address your use case with making `.m` files.

**Comment by janfreyberg** [comment]
@takluyver that seems like a good idea, but I can envision a situation in which I would want more than one option for exporting a file in a particular language. For eg Matlab, `.m` is their basic script extension, but they recently released a proprietary form of notebook-type files (`.mlx` files) which would require a different exporter. So simply overriding the script exporter might be too limiting!

**Comment by takluyver** [comment]
Yep, it doesn't solve the general problem, but it would reduce the pressure for it a little.
